pkg/tun/ip/header.go

```diff
@@ -0,0 +1,132 @@
+package ip
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"net"
+
+	"golang.org/x/net/ipv4"
+	"golang.org/x/net/ipv6"
+)
+
+// A Header provides a common interface for the V4Header and the V6Header
+type Header interface {
+	// Initialize ensures that the header has the correct version and that all other bytes are zero
+	Initialize()
+
+	// Version returns ipv4.Version or ipv6.Version
+	Version() int
+
+	// Destination returns the destination IP
+	Destination() net.IP
+
+	// Source returns the source IP
+	Source() net.IP
+
+	// 	HeaderLength() is the length of this header
+	HeaderLen() int
+
+	// PayloadLen is the length of the payload
+	PayloadLen() int
+
+	// L4Protocol is the protocol of the layer-4 header.
+	L4Protocol() int
+
+	// PacketBase returns the full packet that this header is backed by.
+	Packet() []byte
+
+	// Payload returns the payload of the package that this header is backed by.
+	Payload() []byte
```

Comment:
Since this includes the payload and not just the header, shouldn't this be "ip.Packet", not "ip.Header"?

Reply:
It doesn't really "include" anything. It's just a cast of a `[]byte` that exposes all header fields. I use the term `Packet` for things that holds on to the buffer and have methods to retrieve the IPHeader. That terminology is used throughout so there is also `tcp.Header` and `tcp.Packet` for instance.

Reply:
So I guess I was saying that it's surprising to me that `header.Payload()` is possible; it's surprising that `len(header) != header.HeaderLen()`.